test/TestApps/PushNotificationsTestApp/main.cpp

```diff
@@ -0,0 +1,423 @@
+using namespace winrt::Microsoft::Windows::AppLifecycle;
+using namespace winrt::Microsoft::Windows::PushNotifications;
+using namespace winrt;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::Storage::Streams;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::ApplicationModel::Activation;
+using namespace winrt::Windows::ApplicationModel::Background; // BackgroundTask APIs
+
+enum UnitTest {
+    channelRequestUsingNullRemoteId, channelRequestUsingRemoteId, multipleChannelRequestUsingSameRemoteId,
+    multipleChannelRequestUsingMultipleRemoteId, threeChannelRequestUsingSameRemoteId, activatorTest,
+};
+
+static std::map<std::string, UnitTest> switchMapping;
+winrt::guid remoteId1(L"a2e4a323-b518-4799-9e80-0b37aeb0d225");
+winrt::guid remoteId2(L"CA1A4AB2-AC1D-4EFC-A132-E5A191CA285A");
+winrt::guid remoteId3(L"40FCE789-C6BF-4F47-A6CF-6B9C1DCE31BA");
+
+void signalPhase(const std::wstring& phaseEventName)
+{
+    wil::unique_event phaseEvent;
+    if (phaseEvent.try_open(phaseEventName.c_str(), EVENT_MODIFY_STATE, false))
+    {
+        phaseEvent.SetEvent();
+    }
+}
+
+void initUnitTestMapping()
+{
+    switchMapping["ChannelRequestUsingNullRemoteId"] = UnitTest::channelRequestUsingNullRemoteId;
+    switchMapping["ChannelRequestUsingRemoteId"] = UnitTest::channelRequestUsingRemoteId;
+    switchMapping["MultipleChannelRequestUsingSameRemoteId"] = UnitTest::multipleChannelRequestUsingSameRemoteId;
+    switchMapping["MultipleChannelRequestUsingMultipleRemoteId"] = UnitTest::multipleChannelRequestUsingMultipleRemoteId;
+    switchMapping["ThreeChannelRequestUsingSameRemoteId"] = UnitTest::threeChannelRequestUsingSameRemoteId;
+    switchMapping["ActivatorTest"] = UnitTest::activatorTest;
+}
+
+bool ChannelRequestUsingNullRemoteId()
+{
+    winrt::hresult hr = S_OK;
+
+    try
+    {
+        auto channelOperation = PushNotificationManager::CreateChannelAsync(GUID_NULL).get();
+    }
+    catch (...)
+    {
+        auto channelRequestException = hresult_error(to_hresult());
+        hr = channelRequestException.code();
+    }
+
+    return (hr == E_INVALIDARG);
+}
+
+bool ChannelRequestUsingRemoteId()
+{
+    wil::unique_handle channelEvent = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+    auto channelOperationResult = S_OK;
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(remoteId1);
+
+    // Setup the completed event handler
+    channelOperation.Completed(
+        [&channelEvent, &channelOperationResult](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& sender,
+            AsyncStatus const /* asyncStatus */)
+        {
+            auto result = sender.GetResults();
+            if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
+            {
+                result.Channel().Close(); // Avoids using cached channel everytime the test runs
```

Comment:
What's this about a cached channel?

Reply:
Sure. The first time we request a channel, for a given GUID, from WNS we store that channel locally in database within the notification platform. For test purposes we want to test end-to-end everytime i.e. the channel request has to go to the WNS and return with the channel and give it to the application requesting it. So we are calling close() on the channel after we fetch it.

Reply:
I think it may be best here to remove that comment and not say anything. Close() assumes cleanup